docker_parallel_runner.py

```diff
@@ -0,0 +1,320 @@
+        # Instead we will prepare the SSH key once before starting any suites.
+        return self._run([self._rebuild_script],
+                         extra_env={"SKIP_SSH_KEYGEN": "1"})
+
+    def up(self):
+        return self._run([self._up_script])
+
+    def wait_ready(self):
+        return self._run([self._ready_script])
+
+    def down(self):
+        return self._run([self._down_script])
+
+    def fetch_logs(self):
+        return self._run([self._fetch_logs_script])
+
+    def run_test_suite(self, test_suite: str):
+        return self._run([self._test_suite_script, test_suite])
+
+
+def get_suites(suite_strs: Tuple[str]) -> List[Suite]:
+    suites = []
+
+    # These are used for checking for duplicates
+    suite_files = set()
+    docker_compose_project_names = set()
+
+    for suite_str in suite_strs:
+        logging.info(f"Parsing suite str '{suite_str}'")
+        try:
+            # E.g. "suites/test_something.json:6:2"
+            test_suite, rp_cluster_size, client_count = suite_str.split(":")
+            rp_cluster_size = int(rp_cluster_size)
+            client_count = int(client_count)
+        except Exception:
+            raise ValueError(f"Invalid suite: {suite_str}")
+
+        if (rp_cluster_size, client_count) not in POSSIBLE_SIZE_COMBOS:
+            # Underlying bash scripts literally only support these three combos.
+            # Hence, we explicitly error out on bad input early.
+            raise ValueError(f"Invalid cluster size, client count: ({rp_cluster_size}, {client_count})")
+
+        suite = Suite(suite_file=test_suite,
+                      rp_cluster_size=rp_cluster_size, client_count=client_count)
+
+        if suite.suite_file in suite_files:
+            raise ValueError(f"Duplicate test suite: {suite.suite_file}")
+        suite_files.add(suite.suite_file)
+
+        if not os.path.exists(suite.suite_file):
+            raise ValueError(f"Test suite file not found: {suite.suite_file}")
+
+        if suite.docker_compose_project_name in docker_compose_project_names:
+            raise ValueError(f"Duplicate docker-compose project name: {suite.docker_compose_project_name}")
+        docker_compose_project_names.add(suite.docker_compose_project_name)
+
+        suites.append(suite)
+    return suites
+
+
+def run_suites(suite_strs: Tuple[str], core_count_mult: float):
+    suites = get_suites(suite_strs)
+
+    logging.info(f"Got test suites: {suites}")
+
+    logging.info(
+        "Generating SSH key (if none exists). "
+        "Used by all test suites, so do it once at beginning to avoid race."
+    )
+    generate_ssh_key()
+
+    # We implement our own scheduling based on CPU availability
+    # We leverage the ThreadPoolExecutor for its simple threading interface
+    # Therefore we effectively do not constrain on max_workers (set to theoretical max usage == len(suites))
+    executor = ThreadPoolExecutor(max_workers=len(suites))
+
+    actual_cpu_available = os.cpu_count()
```

Comment:
I think we need to exclude hyperthreading in amd64.  `os.cpu_count()` returns the cpu count without excluding hyperthreading. Example:

```bash
$ cat /proc/cpuinfo | grep core.id | sort | uniq | wc -l
24
$ python3
>>> import os
>>> os.cpu_count()
48
>>>
```

Reply:
We are just trying to see how much "CPU power" there is on the machine.  For that purpose, the logical core count works fine.  Note the RP brokers we are running are in "dev" mode, no CPU pinning involved.  In the end it just boils down to calibration - during testing 1 container per 1 logical CPU core seems to work fine.

Would be interested to hear of other reasons for specifically using physical cores here.

Reply:
I'm reading up to get up to speed on this hyperthreading business.  Is the idea that logical cores are much weaker regular cores, and introduces inconsistent resourcing across ARM / X86?

Reply:
About to merge this - this was a useful callout @gousteris.  IMO, there is some value to maintaining simplicity here, given things are working fine.  I.e. to really account for hyperthreads and real cores, we would:

* count physical cores:
   * 48 for ARM, 24 for X86.
* In our additional app logic, apply a 2x mult to the X86 number to get back to 48.
* Proceed as usual with 48 effective.

We want to maintain effective 48, because using just 24 effective would double run duration, just for x86.

I'm slightly hesitant to add this complication as it's somewhat inconsequential right now (i.e. both ARM / X86 share same code path with same parallelism, and run duration).

I will however add a detailed comment about this consideration.  E.g. may be ARM job can run with even greater parallelism!  This is not impactful right now, because we really only care about the max_duration(arm, x86).

Reply:
main motivation for this comment is that we already do that in CI, ie we exclude hyperthreading when spinning up rp containers